benchmarks/src/main/java/org/opensearch/common/round/RoundableBenchmark.java

```diff
@@ -135,6 +136,19 @@ public void setup() {
                 case "linear":
                     supplier = () -> new BidirectionalLinearSearcher(values, size);
                     break;
+                case "btree":
```

Comment:
I would prefer to replace reflection with `RoundableBenchmark` with `src/main/java20`, much cleaner and straightforward way to run benchmarks, wdyt?

Reply:
I tried this but it gets a little tricky. `:benchmarks` and `:libs:opensearch-common` are separate modules. While the benchmarks can access the `BtreeSearcher` class at runtime, I don't think it's possible to make it visible at compile time.

It was possible to do this within the same module ([example](https://github.com/opensearch-project/OpenSearch/pull/11194/commits/550422cdc25a6263dfa6dadef4803a7daf553203#diff-332bcff66712f49326d714f043d507a7d3a203bd829d00a84be17785f1045342R71-R75)), but I don't think it's possible to do it across different modules. I think the correct way forward is to not have a separate benchmarks module. Relevant benchmarks should live in the same module under a different source, say `<module>/src/jmh`. That way modules are also self-contained. Thoughts? I can create a GitHub issue for this and take up later.

Reply:
It looks it bit tricky but feasible, do you mind I push few changes related to `benchmarks` ?  (it is a bit difficult to comment on changes required).

Reply:
Ah! I saw your diff. I didn't know `project(':libs:opensearch-common').sourceSets.java20.output` could be used this way to access sources from a different module. :D

Reply:
@reta I think this approach still requires some plumbing work. I tried running the benchmarks on top of your changes, but it doesn't run "btree" tests, indicating it's picking up older classes.

First problem is, the multi-release JAR isn't used in the classpath. Here's the output from `ps aux` when running benchmarks.
```
~/.sdkman/candidates/java/21-open/bin/java \
    --add-modules=jdk.incubator.vector \
    -cp /Volumes/workplace/performance/simd/OpenSearch/benchmarks/build/classes/java/main:/Volumes/workplace/performance/simd/OpenSearch/benchmarks/build/resources/main:...truncated...
```

I fixed this by adding the following to the Gradle config.
```
run.classpath += files(jar.archiveFile)
run.classpath -= sourceSets.main.output
```

This fixes the classpath problem, yet, the benchmarks still use the older classes. I believe this is happening due to code generation done by JMH; harness is only being created for the main sources.

Reply:
> This fixes the classpath problem, yet, the benchmarks still use the older classes. I believe this is happening due to code generation done by JMH; harness is only being created for the main sources.

I think there is one part still missed - annotation processing, which apparently is built off the main codebase, I will look today, thanks!

Reply:
@ketanv3 classes are properly processed now (with JMH annotation processor) but there is an issue with JMH metadata itself that is stored in `META-INF/versions/20/META-INF/BenchmarkList`, this is binary file and if I am not mistaken, it has `@Param` expansion there.

Reply:
@reta Yes, I was thinking something similar. But instead of noop, fail explicitly with an ClassNotFoundException since the former would run too fast and produce meaningless results.

You're right about the param expansion in the BenchmarkList file. Each line in that file corresponds to a [BenchmarkListEntry](https://github.com/openjdk/jmh/blob/master/jmh-core/src/main/java/org/openjdk/jmh/runner/BenchmarkListEntry.java#L65) which contains the required params. I'll too take a look at this today and see if there's a clean workaround.

Otherwise, what do you feel about going back to the original approach? Though there's reflection, I found it to be a much cleaner and comprehensible.

Reply:
thanks @ketanv3 

> @reta Yes, I was thinking something similar. But instead of noop, fail explicitly with an ClassNotFoundException since the former would run too fast and produce meaningless results.

It also works

> Otherwise, what do you feel about going back to the original approach? Though there's reflection, I found it to be a much cleaner and comprehensible.

I honestly see no reasons to go to original approach if we add the "btree" to the parameters, the right implementation will be picked up, right?

Reply:
@reta The approach sounds good to me. My only concern is code duplication - the two variants of the same benchmark may diverge in undesired ways.

To minimize code duplication, I've now created a separate class `RoundableSupplier` which gets overridden in Java 20 sources, while the benchmark class remains common to both.

Reply:
Sounds great, thanks @ketanv3 !